### ISSUE TITLE
paging results example

### ISSUE THREAD

**Issue by SogoGolf** [Azure/Azure.Android#82]
hi all, what's the right way to go about paging through a query that returns (say) 1400 documents ?

```
            AzureData.queryDocuments ("xxxxx", "xxxxx", query, ClubDocument::class.java) {
                
                //first set of 100 documents
                val documents = it.resource?.items

                if (it.isSuccessful && it.response != null) {

                    documents?.listIterator()?.forEach { club ->
                        Log.i("AZURE", "${club.name}")
                    }
                    
                    //loop through remaining docs..? doesn't work keeps returning the 2nd page (101-200)
                    while (it.hasMoreResults) {
                        it.next {
                            it.resource?.items?.listIterator()?.forEach { club ->
                                Log.i("AZURE", "${club.name}")
                            }
                        }
                    }

      ......
     ......


```

**Comment by NateRickard** [comment]
I added a paged Query test in #89

I think what may be happening with your code is the `it` you're looping on is captured and never updated.  That is, `while (it.hasMoreResults) {` is checking the same value over and over.  If you introduce a local variable to store the last response it should work, e.g.:

```
var listResponse = it

while (listResponse.hasMoreResults) {

...

it.next {
   listResponse = it
...

```

**Comment by SogoGolf** [comment]
thanks @NateRickard . so i can get the paging to work now but for some reason i can't break out of the while loop once the paging is complete. ```it?.hasMoreResults``` is being set to ```false``` once the cursor is done so all good there. I've tried this lots of different ways, find the pattern pretty confusing. can you help point out what im missing to get paging through this result set working? 


```
val query = Query.select()
    .from("sogocollection")
    .where("type", "golfer")

var waitForResponse: ListResponse<GolferDocument>? = null

AzureData.queryDocuments ("sogocollection", "sogodb", query, GolferDocument::class.java) {

    waitForResponse = it

    while (it?.hasMoreResults!!) {

        waitForResponse?.let { response ->

            waitForResponse = null

            response.next {
                it?.resource?.items?.listIterator()?.forEach { golfer ->
                    Log.i("AZURE", "${golfer.id} ${golfer.firstName} ${golfer.lastName} $")
                }

                waitForResponse = it

                Log.i("AZURE", "batch count done")
                Log.i("AZURE", "**************************************************")
            }
        }
    }



```

**Comment by colbylwilliams** [comment]
so... this isn't Kotlin, or Swift for that matter, but a temporary solution would be to use recursion similar to the following (sloppy sudo-code):

```kotlin
val query = Query.select()
    .from("sogocollection")
    .where("type", "golfer")

var documents = listOf<GolferDocument>()

func getGolferDocuments(passedInResponse: ListResponse<GolferDocument>?) {

    if passedInResponse == null {

        AzureData.queryDocuments ("sogocollection", "sogodb", query, GolferDocument::class.java) {

            documents.add(it?.resource?.items)

            if (it?.hasMoreResults) {
                getGolferDocuments(it)
            }
        }
    } else {

        passedInResponse.next {

            documents.add(it?.resource?.items)

            if (it?.hasMoreResults) {
                getGolferDocuments(it)
            }
        }
    }
}
```

...but there's definitely a gap in the API that we could improve.

**Comment by NateRickard** [comment]
thanks @colbylwilliams ... @SogoGolf I've been chatting with colby on this since the real issue is that the `.next()` call goes and grabs the documents asynchronously on a background thread, which allows the loop to keep going on the current thread, causing a lot of synchronization issues.  His code would in theory clean that up a bit, and we'll look at improving the api.

One thing I'm also doing is providing a Kotlin [coroutine-based](https://kotlinlang.org/docs/reference/coroutines-overview.html) api that would look and feel a little bit more graceful (no callbacks needed).  I have this prototyped for the `next()` function, but unless you're prepared to use coroutines in your app and/or otherwise sync the threading that may not be super easy to grok.

I've added tests for both methods (coroutine & recursion) in #91, so you can see how you could implement something similar.  We'll work on encapsulating the recursion stuff into an api that allows you to get x # of pages/all pages with one call.

**Comment by SogoGolf** [comment]
hey @colbylwilliams @NateRickard thanks very much really appreciated. also ill take a look at the tests you noted above in the mean time.

**Comment by NateRickard** [comment]
@colbylwilliams @SogoGolf  Check out the API improvements in #96.  Ideally I would have liked a fluent API that we could chain to the initial call, but the callbacks kinda screw that up.  This PR would enable you to call `.getMorePages(<#OfPages>)` or `getAllPages()` on a ListResponse and it would return all of those documents - essentially doing all the recursion in the SDK.  Open to improvements if you can spot any ways to make this better.  Also includes tests if you want to see it in action.

**Comment by SogoGolf** [comment]
@NateRickard that sounds like a huge improvement! let me know when i can take these api updates for a spin

**Comment by NateRickard** [comment]
@SogoGolf available to use in 0.2.3 which I just released
